File: 25.BinarySearchOnRotatedArray.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.io.*;

public class Main {

    // function to search in rotated sorted array
    static int searchRotated(int[] nums, int target) {
        int left = 0;                                 // left pointer of search space
        int right = nums.length - 1;                  // right pointer of search space

        while (left <= right) {                       // continue while valid range exists
            int mid = (left + right) / 2;             // middle index of current range

            if (nums[mid] == target) {                // if we found target at mid
                return mid;                           // return index
            }

            if (nums[left] <= nums[mid]) {            // check if left half [left..mid] is sorted
                if (nums[left] <= target && target < nums[mid]) {
                    right = mid - 1;                  // target in left half -> move right
                } else {
                    left = mid + 1;                   // target not in left half -> move left
                }
            } else {                                  // right half [mid..right] is sorted
                if (nums[mid] < target && target <= nums[right]) {
                    left = mid + 1;                   // target in right half -> move left
                } else {
                    right = mid - 1;                  // target not in right half -> move right
                }
            }
        }

        return -1;                                    // target not found
    }

    public static void main(String[] args) throws Exception {
        PrintWriter out = new PrintWriter(System.out);

        int[] arr = {4, 5, 6, 7, 0, 1, 2};              // rotated sorted array example
        out.println("Array: [4, 5, 6, 7, 0, 1, 2]");
        out.println("search 0 -> " + searchRotated(arr, 0));  // expected index 4
        out.println("search 3 -> " + searchRotated(arr, 3));  // expected -1

        int[] arr2 = {1, 2, 3, 4, 5};                   // normal sorted array
        out.println("Array2: [1, 2, 3, 4, 5]");
        out.println("search 4 -> " + searchRotated(arr2, 4)); // expected index 3

        out.flush();
    }
}

File: 25.BinarySearchOnRotatedArray.py
# Binary Search on a rotated array finds a target value in a sorted array 
# that has been rotated (shifted) at some pivot.
# It uses a modified binary search: at each step determine which half is sorted and 
# decide which half can possibly contain the target.

# Explanation

# A rotated sorted array looks like [4,5,6,7,0,1,2] — 
# originally sorted but rotated at pivot between 7 and 0. 
# Standard binary search fails because the whole array isn't strictly sorted. 
# The trick: at every step check whether the left or right half is sorted (compare nums[left] and nums[mid]). 
# If the left half is sorted and the target lies inside that sorted range, search left; 
# otherwise search right. Mirror the logic if the right half is sorted. 
# This keeps the search O(log n) for arrays without duplicates.

def search_rotated(nums, target):
    left = 0 
    right = len(nums) - 1

    while left <= right:
        mid = (left + right) // 2
        if nums[mid] == target:
            return mid
        
        if nums[left] <= nums[mid]:
            if nums[left] <= target < nums[mid]:
                right = mid - 1
            else:
                left = mid + 1

        else: 
            if nums[mid] < target <= nums[right]:
                left = mid + 1
            else:
                right = mid - 1

    return -1

arr = [4, 5, 6, 7, 0, 1, 2]
print("Array:", arr)
print("search 0 ->", search_rotated(arr, 0))
print("search 3 ->", search_rotated(arr, 3))

arr2 = [1, 2, 3, 4, 5]
print("Array2:", arr2)
print("Search 4 ->", search_rotated(arr2, 4))


